File: packages/types/src/scan-result.ts
export interface Technology {
  name: string;
  category: string;
  version?: string;
  confidence: number;
  website?: string;
  icon?: string;
}

export interface SSLInfo {
  valid: boolean;
  issuer: string;
  expires_at: string;
  protocol: string;
  days_until_expiry: number;
}

export interface HeaderCheck {
  name: string;
  present: boolean;
  value?: string;
  recommendation?: string;
}

export interface SecurityScore {
  score: number;
  grade: "A+" | "A" | "B" | "C" | "D" | "F";
  ssl: SSLInfo;
  headers: HeaderCheck[];
  issues: string[];
  https_redirect: boolean;
}

export interface CompanyInfo {
  name?: string;
  description?: string;
  logo?: string;
  industry?: string;
  founded?: string;
  location?: string;
  employees_range?: string;
  type?: string;
}

export interface SocialLink {
  platform: string;
  url: string;
  handle?: string;
  followers?: number;
}

export interface SeoData {
  domain_age_days?: number;
  meta_title?: string;
  meta_title_length?: number;
  meta_description?: string;
  meta_description_length?: number;
  has_sitemap: boolean;
  has_robots: boolean;
  heading_structure: { h1: number; h2: number; h3: number };
  has_structured_data: boolean;
  canonical_url?: string;
  score: number;
}

export interface BusinessSignals {
  has_pricing: boolean;
  has_careers: boolean;
  ad_pixels: string[];
  chat_widgets: string[];
  payment_processors: string[];
  is_hiring: boolean;
  is_monetized: boolean;
  email_providers: string[];
}

export interface TrafficData {
  tranco_rank?: number;
  traffic_tier: "Very High" | "High" | "Medium" | "Low" | "Very Low";
  estimated_monthly_visits?: string;
}

export interface Competitor {
  url: string;
  name?: string;
  similarity_score: number;
  shared_technologies: string[];
}

export interface CostItem {
  category: string;
  service: string;
  min_monthly: number;
  max_monthly: number;
}

export interface CostEstimate {
  total_monthly_min: number;
  total_monthly_max: number;
  breakdown: CostItem[];
  currency: string;
}

export interface CloneInfo {
  is_template: boolean;
  template_name?: string;
  structural_hash: string;
}

export interface DnsInfo {
  a_records: string[];
  aaaa_records: string[];
  mx_records: string[];
  ns_records: string[];
  txt_records: string[];
  hosting_provider?: string;
}

export interface SecurityFix {
  summary: string;
  nginx: string | null;
  apache: string | null;
  cloudflare: string | null;
  meta_tag: string | null;
  vercel_json: string | null;
  netlify_toml: string | null;
}

export interface SecurityFinding {
  id: string;
  severity: "critical" | "high" | "medium" | "low" | "info";
  title: string;
  description: string;
  impact: string;
  fix: SecurityFix;
  effort: "5min" | "15min" | "30min" | "1hr" | "complex";
  priority: number;
}

export interface CarbonScore {
  co2_grams_per_visit: number;
  energy_kwh_per_visit: number;
  is_green_hosted: boolean;
  hosting_provider: string | null;
  sustainability_grade: string;
  page_weight_bytes: number;
  cleaner_than_percent: number;
  annual_co2_kg: number;
  estimated_monthly_visits: number;
  recommendations: string[];
}

export interface EmailPatterns {
  found_emails: string[];
  pattern: string | null;
  confidence: number;
  team_page_url: string | null;
  contact_page_url: string | null;
}

export interface OffensiveFinding {
  id: string;
  severity: "critical" | "high" | "medium" | "low" | "info";
  category: string;
  title: string;
  description: string;
  evidence: string;
  remediation: string;
  cvss_score?: number;
  cwe_id?: string;
  affected_asset: string;
  tool_used: string;
  request_payload?: string;
  response_snippet?: string;
  proof_of_concept?: string;
}

export interface SecretScanResult {
  secrets: {
    type: string;
    value_preview: string;
    source_url: string;
    line?: number;
    severity: "critical" | "high" | "medium" | "low";
  }[];
  total_found: number;
  sources_checked: number;
}

export interface WafResult {
  detected: boolean;
  provider: string | null;
  confidence: number;
  evidence: string[];
}

export interface GraphqlDetectResult {
  detected: boolean;
  endpoint: string | null;
  introspection_enabled: boolean;
  playground_enabled: boolean;
  engine?: string;
}

export interface MobileFinding {
  id: string;
  severity: "critical" | "high" | "medium" | "low" | "info";
  category: string;
  title: string;
  description: string;
  evidence: string;
  remediation: string;
  cvss_score?: number;
  cwe_id?: string;
  affected_asset: string;
  tool_used: string;
  request_payload?: string;
  response_snippet?: string;
  proof_of_concept?: string;
  platform: "android" | "ios" | "cross-platform";
  owasp_mobile_id?: string;
  component?: string;
  appstore_blocker?: boolean;
}

export interface AppMetadata {
  app_name?: string;
  package_name?: string;
  version_name?: string;
  version_code?: string;
  min_sdk?: number;
  target_sdk?: number;
  permissions: string[];
  framework?: string;
  platform: "android" | "ios";
}

export interface MobileScanResult {
  id: string;
  user_id: string;
  app_name?: string;
  package_name?: string;
  platform: "android" | "ios";
  scan_mode: "appstore_check" | "security_audit" | "offensive_pentest";
  file_name: string;
  file_size_bytes: number;
  file_hash_sha256: string;
  status: string;
  findings: MobileFinding[];
  logs: { timestamp: string; level: string; phase: string; message: string }[];
  severity_critical: number;
  severity_high: number;
  severity_medium: number;
  severity_low: number;
  severity_info: number;
  tools_used: string[];
  app_metadata?: AppMetadata;
  api_endpoints_found: string[];
  framework_detected?: string;
  offensive_scan_id?: string;
  started_at?: string;
  completed_at?: string;
  created_at: string;
}

export interface ScanResult {
  id?: string;
  url: string;
  url_hash: string;
  tech_stack: Technology[];
  security: SecurityScore;
  company: CompanyInfo;
  social_links: SocialLink[];
  seo: SeoData;
  business_signals: BusinessSignals;
  traffic: TrafficData;
  competitors: Competitor[];
  cost_estimate: CostEstimate;
  clone_info?: CloneInfo;
  dns?: DnsInfo;
  ai_summary: string;
  scanned_at: string;
  scan_duration_ms?: number;
  carbon?: CarbonScore;
  security_findings?: SecurityFinding[];
  email_patterns?: EmailPatterns;
  offensive_findings?: OffensiveFinding[];
  mobile_findings?: MobileFinding[];
  secrets_passive?: SecretScanResult;
  waf_info?: WafResult;
  graphql_info?: GraphqlDetectResult;
}

File: apps/api/src/models/scan_result.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Technology {
    pub name: String,
    pub category: String,
    pub version: Option<String>,
    pub confidence: f64,
    pub website: Option<String>,
    pub icon: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SSLInfo {
    pub valid: bool,
    pub issuer: String,
    pub expires_at: String,
    pub protocol: String,
    pub days_until_expiry: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HeaderCheck {
    pub name: String,
    pub present: bool,
    pub value: Option<String>,
    pub recommendation: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SecurityScore {
    pub score: u32,
    pub grade: String,
    pub ssl: SSLInfo,
    pub headers: Vec<HeaderCheck>,
    pub issues: Vec<String>,
    pub https_redirect: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CompanyInfo {
    pub name: Option<String>,
    pub description: Option<String>,
    pub logo: Option<String>,
    pub industry: Option<String>,
    pub founded: Option<String>,
    pub location: Option<String>,
    pub employees_range: Option<String>,
    #[serde(rename = "type")]
    pub company_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SocialLink {
    pub platform: String,
    pub url: String,
    pub handle: Option<String>,
    pub followers: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SeoData {
    pub domain_age_days: Option<i64>,
    pub meta_title: Option<String>,
    pub meta_title_length: Option<usize>,
    pub meta_description: Option<String>,
    pub meta_description_length: Option<usize>,
    pub has_sitemap: bool,
    pub has_robots: bool,
    pub heading_structure: HeadingStructure,
    pub has_structured_data: bool,
    pub canonical_url: Option<String>,
    pub score: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HeadingStructure {
    pub h1: u32,
    pub h2: u32,
    pub h3: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BusinessSignals {
    pub has_pricing: bool,
    pub has_careers: bool,
    pub ad_pixels: Vec<String>,
    pub chat_widgets: Vec<String>,
    pub payment_processors: Vec<String>,
    pub is_hiring: bool,
    pub is_monetized: bool,
    pub email_providers: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrafficData {
    pub tranco_rank: Option<u64>,
    pub traffic_tier: String,
    pub estimated_monthly_visits: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Competitor {
    pub url: String,
    pub name: Option<String>,
    pub similarity_score: f64,
    pub shared_technologies: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostItem {
    pub category: String,
    pub service: String,
    pub min_monthly: f64,
    pub max_monthly: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostEstimate {
    pub total_monthly_min: f64,
    pub total_monthly_max: f64,
    pub breakdown: Vec<CostItem>,
    pub currency: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanResult {
    pub id: Option<Uuid>,
    pub url: String,
    pub url_hash: String,
    pub tech_stack: Vec<Technology>,
    pub security: SecurityScore,
    pub company: CompanyInfo,
    pub social_links: Vec<SocialLink>,
    pub seo: SeoData,
    pub business_signals: BusinessSignals,
    pub traffic: TrafficData,
    pub competitors: Vec<Competitor>,
    pub cost_estimate: CostEstimate,
    pub ai_summary: String,
    pub scanned_at: DateTime<Utc>,
    pub scan_duration_ms: Option<u64>,
    // Phase 1: Vulnerability scanner results
    #[serde(skip_serializing_if = "Option::is_none")]
    pub vulnerability: Option<serde_json::Value>,
    // Phase 2: Intelligence results
    #[serde(skip_serializing_if = "Option::is_none")]
    pub privacy: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub wayback: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub js_bundles: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_endpoints: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supply_chain: Option<serde_json::Value>,
    // Phase 6: Performance/Accessibility
    #[serde(skip_serializing_if = "Option::is_none")]
    pub performance: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accessibility: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lighthouse: Option<serde_json::Value>,
    // Phase 7: Sustainability
    #[serde(skip_serializing_if = "Option::is_none")]
    pub carbon: Option<serde_json::Value>,
    // Security findings with actionable fixes
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security_findings: Option<serde_json::Value>,
    // Email pattern detection
    #[serde(skip_serializing_if = "Option::is_none")]
    pub email_patterns: Option<serde_json::Value>,
    // Passive security scanners
    #[serde(skip_serializing_if = "Option::is_none")]
    pub secrets_passive: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub waf_info: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub graphql_info: Option<serde_json::Value>,
    // Phase 9: Vibe-code & BaaS scanners
    #[serde(skip_serializing_if = "Option::is_none")]
    pub firebase_audit: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supabase_audit: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env_leaks: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub llm_keys: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_auth: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub vibe_code: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub nextauth_audit: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub webhook_audit: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub csp_bypass: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_poisoning: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prototype_pollution: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cors_deep: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dependency_audit: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_upload: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub s3_buckets_deep: Option<serde_json::Value>,
    // Viewport screenshot (base64 PNG)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub screenshot: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanRequest {
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanResponse {
    pub scan_id: Uuid,
    pub status: String,
    pub result: Option<ScanResult>,
    pub cached: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ScanRow {
    pub id: Uuid,
    pub url: String,
    pub url_hash: String,
    pub result: serde_json::Value,
    pub scanned_at: DateTime<Utc>,
    pub expires_at: DateTime<Utc>,
}
